src/main/java/com/vmware/vipclient/i18n/util/LocaleUtility.java

```diff
@@ -19,7 +19,8 @@
     
     // A locale fallback priority queue. For now, it only contains the default locale and the source messages (from "messages_source.json").
     private static List<Locale> fallbackLocales = new LinkedList<Locale>(Arrays.asList(getDefaultLocale(), Locale.forLanguageTag(ConstantsKeys.SOURCE)));
-    
+    private static List<Locale> l2FallbackLocales = new LinkedList<Locale>(Arrays.asList(getDefaultLocale(), getSourceLocale()));
```

Comment:
We do not need getSourceLocale() for L2. It is specifically for L3 purposes. "Source" means the locale of the original messages that were translated in multiple languages. It has nothing to do with L2.

Note: The method getSourceLocale must actually be updated to return the locale used in messages_source.json, instead of manually configuring it in vipconfig.properties.  -- This can be done in a future enhancement.

We only add "source" in the fallbackLocales queue because it means messages_source.json is bundled with the application code. This "messages_source,json"  is used only when it has new/updated L3 strings that are not yet in in the actual messages_xx.json of the source locale (a.k.a out of sync). This works for L2 just fine: L2 workflow must find a fallback locale in the queue before it reach "source". If it reaches "source", then it will fail also which is expected.

That being said, why do we need to maintain 2 separate queues? Aside form "source", will the content be  any different? It won't be good to have to maintain 2 queues.

Reply:
I think source can be removed from the fallbacklocales because it's very special. Therer is no source in singleton service, even in L3 flow, many places need to handle it specially.
So remove it from fallbacklocales and when there is no any translation for the requested locale, get 'source' from local.

Reply:
I agree with xiaochao to remove source locale from locale fallback chain, especially for that of L2. 

If reuse the same locale fallback chain with L3, it may confuse/mislead other team members when they need maintain the code and look at the code at first insight that L2 need fallback to source locale.

Since L2 needn't fallback to source locale in theory, why keep the logic?

In addition, if keep fallback to source locale, in debug mode, the log will print message can't find source pattern which may mislead others the error is lack of source pattern file, actually source pattern is not needed for L2.

So I think we should either remove source locale from fallback queue if L2 and L3 share the same queue or keep standalone fallback queue for L2.

Reply:
Regarding Xiaochao's comment:
First, there is no more discussion whether falling back to source messages is necessary for L3. It is an existing feature, even in the old client workflow, and we will continue to support it. The client library must return source message if source message has been recently updated (does not match the message in the messages_<locale>.json). We call this use case the "out-of-sync" stage where the source bundle does not match the translation bundles yet as the translation bundles are lagging behind.

Second, it is not true that source messages does not exist in the service. They do exist, we call them "latest". We actually want to use this especially when source bundle does not exist locally (i.e. in the case of microservices where source messages are not with the microservice that use it. e.g. VCF, Magna)

Third, let's not keep special logic for messages_source. The same fallback logic that we use for all locales apply to it—get from service if service is available, get from local if service is unavailable. To keep it simple for the users, we use the same "offlineResourcesBaseUrl" for local messages_source.json as well. Why keep the logic for "source" separate if it just means another type of bundle?

Reply:
Regarding Huihui's comment (L2):

In debug mode, you should not see a log that says "can't find "source" locale". As I said, the code will iterate over the queue, and will not even reach "source" because it should find the fallback that it needs before reaching "source".  However, even if you see "can't find "xx" locale" in the log, that should be okay. It just means that the expected fallback locale does not exist, which can happen not just for "source" but for any other locale. It is not a bug.

If it does reach source, that is also okay. Imagine that in the future, we can actually enable a feature where the product team can provide a custom "source" data for L2. For now, "source" can be there with no harm.